text/0000-io-safety.md

```diff
@@ -0,0 +1,429 @@
+comments:
+
+```rust
+/// # Safety
+///
+/// `MyType` wraps a `std::fs::File` which handles the low-level details, and
+/// doesn't have a way to reassign or independently drop it.
+unsafe impl IoSafe for MyType {}
+```
+
+It requires `unsafe`, to require the code to explicitly commit to upholding I/O
+safety. With `IoSafe`, the `do_some_io` example should simply add a
+`+ IoSafe` to provide I/O safety:
+
+```rust
+pub fn do_some_io<FD: AsRawFd + IoSafe>(input: &FD) -> io::Result<()> {
+    some_syscall(input.as_raw_fd())
+}
+```
+
+## Gradual adoption
+
+I/O safety and `IoSafe` wouldn't need to be adopted immediately, adoption
+could be gradual:
+
+ - First, `std` adds `IoSafe` with impls for all the relevant `std` types.
+   This is a backwards-compatible change.
+
+ - After that, crates could implement `IoSafe` for their own types. These
+   changes would be small and semver-compatible, without special coordination.
+
+ - Once the standard library and enough popular crates utilize `IoSafe`,
+   crates could start to add `+ IoSafe` bounds (or adding `unsafe`), at their
+   own pace. These would be semver-incompatible changes, though most users of
+   APIs adding `+ IoSafe` wouldn't need any changes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The I/O safety concept
+
+In addition to the Rust language's memory safety, Rust's standard library also
+guarantees I/O safety. An I/O operation is *valid* if the raw handles
+([`RawFd`], [`RawHandle`], and [`RawSocket`]) it operates on are values
+explicitly returned from the OS, and the operation occurs within the lifetime
+the OS associates with them. Rust code has *I/O safety* if it's not possible
+for that code to cause invalid I/O operations.
+
+While some OS's document their file descriptor allocation algorithms, a handle
+value predicted with knowledge of these algorithms isn't considered "explicitly
+returned from the OS".
+
+Functions accepting arbitrary raw I/O handle values ([`RawFd`], [`RawHandle`],
+or [`RawSocket`]) should be `unsafe` if they can lead to any I/O being
+performed on those handles through safe APIs.
+
+Functions accepting types implementing
+[`AsRawFd`]/[`IntoRawFd`]/[`AsRawHandle`]/[`AsRawSocket`]/[`IntoRawHandle`]/[`IntoRawSocket`]
+should add a `+ IoSafe` bound if they do I/O with the returned raw handle.
+
+## The `IoSafe` trait
+
+Types implementing `IoSafe` guarantee that they uphold I/O safety. They must
+not make it possible to write a safe function which can perform invalid I/O
+operations, and:
+
+ - A type implementing `AsRaw* + IoSafe` means its `as_raw_*` function returns
+   a handle which is valid to use for the duration of the `&self` reference.
+   If such types have methods to close or reassign the handle without
+   dropping the whole object, they must document the conditions under which
+   existing raw handle values remain valid to use.
```

Comment:
Could you give an example of a type, other than RawFd or a constructed equivalent, which would not satisfy this property?

Reply:
I'm thinking about cases like this:
```rust
struct Hazardous {
    file: RefCell<std::fs::File>
}

impl Hazardous {
    fn set_file(new_file: std::fs::File) {
        *self.file.borrow_mut() = new_file;
    }
}

impl AsRawFd for Hazardous {
    fn as_raw_fd(&self) -> RawFd {
        self.file.borrow().as_raw_fd()
    }
}
```
If someone calls `set_file` while holding a `RawFd` returned from `as_raw_fd`, the original file may be dropped and the `RawFd` may dangle.

Reply:
@sunfishcode That's a great example, thank you! I can easily imagine someone writing code like that as part of, say, a dynamically reassignable "standard output" object. Would you consider capturing that (perhaps along with a motivating example like that) in the RFC?

Reply:
Sure! I've now added an example of this in 87db7d124dafd3fdbcb407247ac80bc9cf4b2f2a.

Reply:
> If such types have methods to close or reassign the handle without
>  dropping the whole object, they must document the conditions under which
 > existing raw handle values remain valid to use.
 
That sounds useful, but we do not require the same for raw pointers, do we? This amounts to a mandate that for every type that has a method that returns a raw pointer, it must then for each of its other methods document whether they existing raw pointers remain valid to use.

Reply:
With the API revision to use `OwnedFd`/`BorrowedFd`/etc., this whole topic becomes a non-issue. If you have a `RefCell<OwnedFd>`, you can't borrow a `BorrowedFd` in a way that would be live across a mutation through a mutable borrow of the `OnwedFd`. It just falls out from Rust's existing lifetime rules.